File: ai_cacher/demo_cost_comparison.py
from typing import Any, Dict, List

from ai_cacher import AICacherWrapper, SemanticCacheRouter


class MockOpenAIChat:
    def __init__(self):
        self.call_count = 0

    def create(self, messages: List[Dict[str, str]], **kwargs: Any) -> Dict[str, Any]:
        self.call_count += 1
        user_prompt = "".join([m.get("content", "") for m in messages if m.get("role") == "user"])
        response_text = self._generate_response(user_prompt)
        return {
            "choices": [
                {
                    "message": {
                        "role": "assistant",
                        "content": response_text,
                    }
                }
            ],
            "usage": {
                "prompt_tokens": self._count_tokens(user_prompt),
                "completion_tokens": self._count_tokens(response_text),
            },
        }

    @staticmethod
    def _generate_response(prompt: str) -> str:
        return f"AI response for prompt: {prompt[:120]}"

    @staticmethod
    def _count_tokens(text: str) -> int:
        return max(1, len(text.split()))


class MockOpenAIClient:
    def __init__(self):
        self.chat = type("obj", (), {"completions": MockOpenAIChat()})


def count_tokens(payload: Dict[str, Any]) -> int:
    usage = payload.get("usage", {})
    return int(usage.get("prompt_tokens", 0)) + int(usage.get("completion_tokens", 0))


def run_direct(client: MockOpenAIClient, messages: List[Dict[str, str]]) -> Dict[str, Any]:
    response = client.chat.completions.create(messages=messages)
    response["cached_by"] = "Direct-Cloud"
    response["latency_ms"] = 1500
    return response


def run_cached(wrapper: AICacherWrapper, messages: List[Dict[str, str]]) -> Dict[str, Any]:
    return wrapper.create_completion(messages=messages)


def main() -> None:
    messages_batch = [
        [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Summarize the monthly sales report."},
        ],
        [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Summarize the monthly sales report."},
        ],
        [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Convert this list to JSON: apple, banana, cherry."},
        ],
        [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Summarize the monthly sales report."},
        ],
        [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Convert this list to JSON: apple, banana, cherry."},
        ],
    ]

    mock_client = MockOpenAIClient()
    router = SemanticCacheRouter(threshold=0.88)
    cached_wrapper = AICacherWrapper(mock_client, router)

    direct_tokens = 0
    cached_cloud_tokens = 0
    cached_hits = 0
    direct_calls = 0
    cached_calls = 0

    print("=== Direct cloud usage ===")
    for messages in messages_batch:
        result = run_direct(mock_client, messages)
        direct_calls += 1
        tokens = count_tokens(result)
        direct_tokens += tokens
        print(f"Call {direct_calls}: prompt='{messages[-1]['content']}' tokens={tokens}")

    print("\n=== Cached usage ===")
    for messages in messages_batch:
        result = run_cached(cached_wrapper, messages)
        if result.get("cached_by") == "AI-Cacher-Local":
            cached_hits += 1
            tokens = 0
        else:
            cached_calls += 1
            tokens = count_tokens(result)
            cached_cloud_tokens += tokens
        print(
            f"Request: '{messages[-1]['content']}' cached_by={result.get('cached_by')} tokens={tokens}")

    print("\n=== Summary ===")
    print(f"Direct cloud calls: {direct_calls}")
    print(f"Cached cloud calls: {cached_calls}")
    print(f"Local cache hits: {cached_hits}")
    print(f"Total direct tokens: {direct_tokens}")
    print(f"Total cached cloud tokens: {cached_cloud_tokens}")
    print(f"Token reduction: {direct_tokens - cached_cloud_tokens} tokens")
    print(
        f"Estimated cost savings: ${(direct_tokens - cached_cloud_tokens) * 0.000002:.6f} "
        "(assuming $0.002 per 1,000 tokens)"
    )


if __name__ == "__main__":
    main()

File: ai_cacher_angular/src/demo_cost_comparison.ts
// Idea and Co-author: Tim Nguyen (timothynnguyen9@gmail.com)

import { AICacherWrapper } from "./client";
import { SemanticCacheRouter } from "./router";
import { ChatMessage } from "./client";

class MockOpenAIClient {
  chat = {
    completions: {
      create: async ({ messages }: { messages: ChatMessage[] }) => {
        const prompt = messages
          .filter((m) => m.role === "user")
          .map((m) => m.content)
          .join(" ");
        const responseText = `AI response for prompt: ${prompt.slice(0, 120)}`;
        const promptTokens = Math.max(1, prompt.split(/\s+/).filter(Boolean).length);
        const completionTokens = Math.max(1, responseText.split(/\s+/).filter(Boolean).length);
        return {
          choices: [
            {
              message: {
                role: "assistant",
                content: responseText,
              },
            },
          ],
          usage: {
            prompt_tokens: promptTokens,
            completion_tokens: completionTokens,
          },
        };
      },
    },
  };
}

type DemoResult = {
  cached_by?: string;
  latency_ms?: number;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
  };
  [key: string]: unknown;
};

function countTokens(payload: DemoResult): number {
  const usage = payload.usage;
  return (usage?.prompt_tokens ?? 0) + (usage?.completion_tokens ?? 0);
}

async function runDirect(client: MockOpenAIClient, messages: ChatMessage[]): Promise<DemoResult> {
  const response = await client.chat.completions.create({ messages });
  return {
    ...response,
    cached_by: "Direct-Cloud",
    latency_ms: 1500,
  };
}

async function runCached(wrapper: AICacherWrapper, messages: ChatMessage[]): Promise<DemoResult> {
  return wrapper.createCompletion(messages);
}

async function main(): Promise<void> {
  const messagesBatch: ChatMessage[][] = [
    [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "Summarize the monthly sales report." },
    ],
    [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "Summarize the monthly sales report." },
    ],
    [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "Convert this list to JSON: apple, banana, cherry." },
    ],
    [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "Summarize the monthly sales report." },
    ],
    [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "Convert this list to JSON: apple, banana, cherry." },
    ],
  ];

  const mockClient = new MockOpenAIClient();
  const router = new SemanticCacheRouter(0.88);
  const wrapper = new AICacherWrapper(mockClient, router);

  let directTokens = 0;
  let cachedCloudTokens = 0;
  let cachedHits = 0;
  let directCalls = 0;
  let cachedCalls = 0;

  console.log("=== Direct cloud usage ===");
  for (const messages of messagesBatch) {
    const result = await runDirect(mockClient, messages);
    directCalls += 1;
    const tokens = countTokens(result);
    directTokens += tokens;
    console.log(`Call ${directCalls}: prompt='${messages[messages.length - 1].content}' tokens=${tokens}`);
  }

  console.log("\n=== Cached usage ===");
  for (const messages of messagesBatch) {
    const result = await runCached(wrapper, messages);
    if (result.cached_by === "AI-Cacher-Local") {
      cachedHits += 1;
    } else {
      cachedCalls += 1;
      cachedCloudTokens += countTokens(result);
    }
    console.log(
      `Request: '${messages[messages.length - 1].content}' cached_by=${result.cached_by} tokens=${
        result.cached_by === "AI-Cacher-Local" ? 0 : countTokens(result)
      }`,
    );
  }

  console.log("\n=== Summary ===");
  console.log(`Direct cloud calls: ${directCalls}`);
  console.log(`Cached cloud calls: ${cachedCalls}`);
  console.log(`Local cache hits: ${cachedHits}`);
  console.log(`Total direct tokens: ${directTokens}`);
  console.log(`Total cached cloud tokens: ${cachedCloudTokens}`);
  console.log(`Token reduction: ${directTokens - cachedCloudTokens} tokens`);
  console.log(
    `Estimated cost savings: $${((directTokens - cachedCloudTokens) * 0.000002).toFixed(6)} (assuming $0.002 per 1,000 tokens)`,
  );
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
